Add an explainer for subresource loading (#526) (#565)
WICG/webpackage: explainers/subresource-loading.md

# Explainer: Subresource loading with Web Bundles

Last updated: March 2020

We propose a new approach to load a large number of resources efficiently using
a format that allows multiple resources to be bundled, e.g.
[Web Bundles](https://web.dev/web-bundles/).

## Backgrounds

- Loading many unbundled resources is still slower in 2020. We concluded that
  [bundling was necessary in 2018](https://v8.dev/features/modules#bundle), and
  our latest local measurement still suggests that.

> - Caching stories with JS bundlers (e.g. webpack) is tough. They are pretty good

Reviewer: ```suggestion
- The caching story with JS bundlers (e.g. webpack) is tough. They are pretty good
```

Or

```suggestion
- The output of JS bundlers (e.g. webpack) doesn't interact well with the HTTP cache. They are pretty good
```

Author: Thanks! The latter sounds good to me.

  tools but configuring them to work in an optimal way is tough, and sometimes

>   also conflicts with new requirements like dynamic bundling (e.g. small edit

Reviewer: ```suggestion
  they're also incompatible with new requirements like dynamic bundling (e.g. small edit
```

And please link "dynamic bundling" to a description.

Author: Thanks. Done.

Author: Ah, I didn't add link "dynamic bundling". I think it's still an early idea, and we don't have a good description of that in public,

Reviewer 2: This one, maybe?

https://github.com/azukaru/progressive-fetching/blob/master/docs/dynamic-bundling/index.md

Author: Thanks! Let me add a link to that. I think that is okay for us.

  with tree shaking could invalidate everything).

- With JS bundlers, execution needs to wait for the full bytes to come. Ideally
  loading multiple subresources should be able to utilize full streaming and

>   parallelization, while it's not possible if all resources are bundled as one

Reviewer: ```suggestion
  parallelization, but that's not possible if all resources are bundled as one
```

Author: Done. Thanks!

  javascript. (For JS modules execution still needs to be waited for the entire
  tree due to the current
  [deterministic execution model](https://docs.google.com/document/d/1MJK0zigKbH4WFCKcHsWwFAzpU_DZppEAOpYJlIW7M7E/edit#heading=h.5652gd5ks5id))

- Related issues: [#411](https://github.com/WICG/webpackage/issues/411),
  [#526](https://github.com/WICG/webpackage/issues/526)

## Proposal

We propose a new `link rel=”webbundle”` notion [**1**] that lets the browser
fetch bundled resources and attach them to the page. Such as:

```html
<link
  rel="webbundle"
  href="https://example.com/dir/subresources.wbn"
  scope="https://example.com/scope/"
/>
```

If the main document has such a `link` element, a browser must do:

- A browser must fetch the specified Web Bundle, `subresources.wbn`.

- All the subresources whose URLs are under the scope

>   `https://example.com/scope/` must [**2**] be loaded from the specified Web

Reviewer: The primary requirement to avoid fetching the same bytes twice is that "If a subresource under that scope is needed later in the document, that later fetch should block until at least the index of the bundle has downloaded to see if it's there."

It seems secondary to then say that if the subresource within that scope isn't in the bundle, its fetch should fail or otherwise notify the developer: that just prevents delays in starting the subresource fetch.

Author: Thanks! Let me add these almost "as-is".  I think that would help the reader of this explainer to understand the possible behaviors we are thinking in more details.

[Fetch maps proposal](https://discourse.wicg.io/t/proposal-fetch-maps/4259)
might be able to extend to this case.

[**2**] The exact behavior is TBD. There are several possible behaviors when the
bundle does not contain the subresource; e.g. just fail, fallback to the
network, or allow users to specify its behavior with a `link` element's
attribute.

## Example

### The bundle

Suppose that the bundle, `subresources.wbn`, includes the following resources:

```
- https://example.com/scope/a.js (which depends on ./b.js)
- https://example.com/scope/b.js
- https://example.com/scope/c.png
- … (omitted)
```

### The main document

```html
<link rel="webbundle"
    href="https://example.com/dir/subresources.wbn"
    scope="https://example.com/scope/"
/>

<script type=”module” src=”https://example.com/scope/a.js”></script>
<img src=https://example.com/scope/c.png>
```

Then, a browser must fetch the bundle, `subresources.wbn`, and load
subresources, `a.js`, `b.js`, and `c.png`, from the bundle.

## Subsequent loading and Caching

On the subsequent navigations for the same page/origin, it’s desirable that the
resources that are already fetched in the bundle shouldn't be refetched. We're

> thinking about two possibilities here:

Reviewer 3: Both of those option sound like exclusion-based fetching. What we've observed while working on a similar system is that there are scale boundaries where it stops working. When the number of individual items is high, it breaks the length limits for URLs (1) and a digest becomes increasingly unreliable (2).

The third solution is to have an official way to batch resources by inclusion instead. It requires an "inclusion digest" or "inclusion list" instead of an "cached/exclusion digest" though.

Reviewer 2: @jkrems
Thanks! I agree that especially for dynamic bundling it seems to need some "inclusion signal" in requesting phase. I suspect it still could be benefited if it's combined with cache/exclusion approach for better cache efficiency?  A straw-idea could be:

1. The bundle request should express some hash or list that indicates the set of resources it wants.  Possibly the server and client can share a logic to compute the hash / variations for the session to determine what resources are needed.
2. Platform also attaches some digest that indicates exclusive-fetching information for caching.

This explainer didn't tell anything about 1. but probably it'd be better to mention it too. Does it make better sense, or I might be still missing something?

Reviewer 3: Makes sense, I think you covered what I had in mind! And this definitely doesn't have to block this initial PR, more a comment towards future iteration.

1. Embed the details of version / cached resources in the URL
2. Use a cache digest or something (and also embed it in the request)

> Then the server should be able to dynamically subset the Bundle and serve it.

Reviewer: A patch format also seems plausible. If the server looks at the If-None-Match request header to learn what version the client already has cached, it can send down a different response type to patch the bundle instead of sending down the whole bundle. I don't know of any existing systems that do that, so we should talk to more people if that direction sounds good. Roberto Peon has argued for patches.

Author: Thanks! Yes, we are currently discussing this topic actively in other places. I think we'll cover this topic in another explainer!

# Explainer: Subresource loading with Web Bundles

Last updated: March 2020

We propose a new approach to load a large number of resources efficiently using

> a format that allows multiple resources be be bundled, e.g.

Author: Ah, thanks! This is typo. "allows multiple resources *to* be bundled" might be better, I think.
I'll fix that.

[Web Bundles](https://web.dev/web-bundles/).

## Backgrounds

- Loading many unbundled resources is still slower in 2020. We concluded that
  [bundling was necessary in 2018](https://v8.dev/features/modules#bundle), and
  our latest local measurement still suggests that.

- Caching stories with JS bundlers (e.g. webpack) is tough. They are pretty good
  tools but configuring them to work in an optimal way is tough, and sometimes
  also conflicts with new requirements like dynamic bundling (e.g. small edit
  with tree shaking could invalidate everything).

> - With JS bundlers, parsing and execution needs to wait for the full bytes to

Reviewer 4: I think this is not true for parsing.

Chrome is already doing streaming parsing / compiling for classic scripts (probably scaling it to more than one processor is hard, though), and for module scripts each module is parsed / compiled as soon as it is downloaded, without waiting for the entire module tree.

Reviewer 2: Would just dropping 'parsing and' make sense?

Author: Thanks. Let's drop the parsing.

  come. Ideally loading multiple subresources should be able to utilize full
  streaming and parallelization, while it's not possible if all resources are
  bundled as one javascript. (For JS modules execution still needs to be waited
  for the entire tree due to the current
  [deterministic execution model](https://docs.google.com/document/d/1MJK0zigKbH4WFCKcHsWwFAzpU_DZppEAOpYJlIW7M7E/edit#heading=h.5652gd5ks5id))

- Related issues: [#411](https://github.com/WICG/webpackage/issues/411),
  [#526](https://github.com/WICG/webpackage/issues/526)

## Proposal

We propose a new `link rel=”webbundle”` notion [**1**] that lets the browser
fetch bundled resources and attach them to the page. Such as:

```html
<link
  rel="webbundle"
  href="https://example.com/dir/subresources.wbn"
  scope="https://example.org/scope/"
/>
```

> Both the `href` and the `scope` need to be the same-origin.

Reviewer 5: I don't think they need to be the same-origin, and actually the example URLs are not the same-origin.
When I wrote [this example](https://docs.google.com/document/d/18fFrURT6xD1GnqJwCx2K3Z81Te9Iv-r2ZsCkOvDH_04/edit#heading=h.irqypigubog7), my idea was:
  - If the page is controlled by a service worker, the service worker can intercept any requests from the page and can respond with arbitrary resources.
  - So, it is safe (=no difference from service worker from the security point of view) even if they are not the same origin, when the page can access the web bundle in the CORS manner.

Author: Thanks.
@kinu, what do you think? I'm afraid that I misunderstood your suggestions.

I'm wondering what the explainer should mention here. We have 3 URLs:

1. The main document's URL
2. href
3. scope

I thought 2 and 3 need to be the same origin, however, it is okay that 1 is in a different origin.
It's a simple mistake I forgot to update example URLs.

Any suggestions? If we don't have a clear answer, I think it is okay to mention it as TBD.

Reviewer 2: Yeah I suggested this but I remember I discussed that cross-origin seems okay too with you. I want to think through all the implications... for example it apparently complicates CSP story. But I agree that maybe we can just drop this for now.

Reviewer 2: (^^I commented before I see Hayato's response) @hayatoito I was indeed thinking about 2 and 3 indeed, but thinking about this more it's a little unclear too. (As @horo-t mentioned SW could already override it)  I think I'm fine with just dropping this sentence.

Author: Thanks! Okay, let's just drop this for now.
Once we have a more clear idea, let's add something back.

If the main document has such a `link` element, a browser must do:

- A browser must fetch the specified Web Bundle, `subresources.wbn`.

- All the subresources whose URLs are under the scope
  `https://example.org/scope/` must [**2**] be loaded from the specified Web
  Bundle.

[**1**] The syntax is tentative. The API surface is not so important for us yet.
Any alternative ideas are welcome. This is basically about mapping a set of

> resources to a particular bundled resource. Something like Fetch maps proposal

Reviewer 4: ```suggestion
resources to a particular bundled resource. Something like [Fetch maps proposal](https://discourse.wicg.io/t/proposal-fetch-maps/4259)
```

Author: Thanks! Let me add a link.

might be able to extend to this case.

[**2**] The exact behavior is TBD. There are several possible behaviors when the
bundle does not contain the subresource; e.g. just fail, fallback to the
network, or allow users to specify its behavior with a `link` element's
attribute.

## Example

### The bundle

Suppose that the bundle, `subresources.wbn`, includes the following resources:

```
- https://example.org/scope/a.js (which depends on ./b.js)
- https://example.org/scope/b.js
- https://example.org/scope/c.png
- … (omitted)
```

### The main document

```html
<link rel="webbundle"
    href="https://example.com/dir/subresources.wbn"
    scope="https://example.org/scope/"
/>

<script type=”module” src=”https://example.com/scope/a.js”></script>
<img src=https://example.org/scope/c.png>
```

Then, a browser must fetch the bundle, `subresources.wbn`, and load
subresources, `a.js`, `b.js`, and `c.png`, from the bundle.

## Subsequent loading and Caching

On the subsequent navigations for the same page/origin, it’s desirable that the
resources that are already fetched in the bundle shouldn't be refetched. We're
thinking about two possibilities here:

> 1. Embed the details of version / cached resources in the URL

Reviewer 4: Could you expand this a bit? I'm not sure how this enables resource subsetting.

Author: @kinu 
Do you have any opinions? I guess we don't have a concrete idea yet.

Author: Can we address this in a follow-up commit? We can adress that later, I think.

Reviewer 4: sgtm.

Reviewer 2: Yeah... I think I was thinking *something* like: Server can maintain versioned bundles, and client can send a request with some URL parameters that indicates the version of the bundle they already have in cache (so that the server can identify the updated resources and send them back).  This is mostly about updating-fetch for same bundle cases though, and haven't really thought about how to do for similarly looking bundle on different pages cases.

Author: @kinu Thanks! I appreciate that! That's helpful. I think we can update the explainer with the more details once we can explore this area. Let me close this as resolved for now.
